### PR TITLE
Add numerical equivalence check to GraphOptzTest.BatchNormAfterConvNotOptimizeForTrain

### PR DESCRIPTION
**Pull request by allibell**
Summary:
Add a numeric equivalence test to the `batchNormAfterConvNotOptimizeForTrain` subtest in `GraphOptzTest`.  The original test ensures there are no structural changes to the graph when `BatchNormalization` nodes are optimized/compiled in training mode; this addition confirms that there is *numeric* equivalence on the original and optimized graphs.

Test Plan:
`ninja all && test/GraphOptzTest`
```
[----------] Global test environment tear-down
[==========] 128 tests from 4 test suites ran. (977 ms total)
[  PASSED  ] 128 tests.
```

### CODE REVIEW

File: tests/unittests/GraphOptzTest.cpp [hunk 1]
```diff
@@ -726,10 +726,10 @@ TEST_F(GraphOptz, BatchNormAfterConvNotOptimizeForTrain) {
 
   EXPECT_EQ(F_->getNodes().size(), 3);
 
-  ::glow::optimize(F_, CompilationMode::Train);
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: This is a training-specific use case so unfortunately I think we need to leave this in and clone the graph manually.

File: tests/unittests/GraphOptzTest.cpp [hunk 2]
```diff
@@ -726,10 +726,14 @@ TEST_F(GraphOptz, BatchNormAfterConvNotOptimizeForTrain) {
 
   EXPECT_EQ(F_->getNodes().size(), 3);
 
+  auto *clone = F_->clone(F_->getName().str() + "_clone");
+  ::glow::convertPlaceholdersToConstants(F_, bindings_, {});
+  optimizedF_ = optimizeFunction(clone);
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]: We need to optimize `optimizedF_` for training.
```suggestion
::glow::optimize(optimizedF_, CompilationMode::Train);
```

File: tests/unittests/GraphOptzTest.cpp [hunk 3]
```diff
@@ -726,10 +726,14 @@ TEST_F(GraphOptz, BatchNormAfterConvNotOptimizeForTrain) {
 
   EXPECT_EQ(F_->getNodes().size(), 3);
 
+  auto *clone = F_->clone(F_->getName().str() + "_clone");
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]:
```suggestion
optimizedF_ = F_->clone(F_->getName().str() + "_clone");
```

File: tests/unittests/GraphOptzTest.cpp [hunk 4]
```diff
@@ -726,10 +726,14 @@ TEST_F(GraphOptz, BatchNormAfterConvNotOptimizeForTrain) {
...
+  ::glow::convertPlaceholdersToConstants(F_, bindings_, {});
+  optimizedF_ = optimizeFunction(clone);
+  EXPECT_EQ(optimizedF_->getNodes().size(), 3);
+
   ::glow::optimize(F_, CompilationMode::Train);
```
Comment (SplitInfinity) [inline tests/unittests/GraphOptzTest.cpp]:
```suggestion
```
We want to leave `F_` unoptimized.

File: tests/unittests/GraphOptzTest.cpp [hunk 5]
```diff
@@ -726,10 +726,14 @@ TEST_F(GraphOptz, BatchNormAfterConvNotOptimizeForTrain) {
 
   EXPECT_EQ(F_->getNodes().size(), 3);
 
+  auto *clone = F_->clone(F_->getName().str() + "_clone");
+  ::glow::convertPlaceholdersToConstants(F_, bindings_, {});
```
[the remark this answers is not in the document]
Comment (allibell) [reply tests/unittests/GraphOptzTest.cpp]: Sorry, my n00by mistake, didn't mean to include this.

**Comment by allibell** [comment]
Hmm.  Oddly, they look the same.  Here are the images: [optimized](https://user-images.githubusercontent.com/25647843/75928462-4d2ecb00-5e23-11ea-93bd-165571586481.png) and [original](https://user-images.githubusercontent.com/25647843/75928561-86ffd180-5e23-11ea-8b2b-032b0db61826.png).  Diff-ing the files confirms that they're the same.  Here is `F_` before my changes: [before](https://user-images.githubusercontent.com/25647843/75929414-6cc6f300-5e25-11ea-97f0-6af99bc9425f.png)

I apologize for my currently limited knowledge of the codebase, but I'm investigating now.

**Comment by SplitInfinity** [comment]
Okay, so the reason they look so different is because the first two dot graphs you linked have already been lowered. However, I think I made a mistake in my earlier statement. This optimization is testing that nothing happens when a `BatchNormalization` node is compiled in training mode, so `F_` and `optimizedF_` should actually look the same.

**Comment by SplitInfinity** [comment]
Okay, everything looks good, but before I sign off on this, can you squash all of the commits into one?

**Comment by allibell** [comment]
Okay, that makes sense.  Squashing now.  Thanks so much for your help!

**Comment by SplitInfinity** [comment]
Almost there! Can you fix the commit message so that it is more descriptive? It seems like it is a vestige from one of the three commits you used to have.

**Comment by allibell** [comment]
Let me know if I can still clarify anything. 🙂

**Comment by SplitInfinity** [comment]
Commit message still looks stale but PR message looks good!

**Pull request closed without merging**